team_docs/samples.md

```diff
@@ -0,0 +1,239 @@
+
+The action performed by an effectful function must be accompanied by a
+"before" vs "after" comparison, where it should be clear what happened.
+
+```kotlin
+val list = mutableListOf(4, 4, 1, 3, 2, 5, 0)
+// list == [4, 4, 1, 3, 2, 5, 0]
+list.sort()
+// list == [0, 1, 2, 3, 4, 4, 5]
+```
+
+We need duplicated elements to demonstrate that they are not merged into one.
+
+```kotlin
+val list = mutableListOf(4, 4, 1, 3, 2, 5, 0)
+// list == [4, 4, 1, 3, 2, 5, 0]
+list.removeAt(0) // 4
+// list == [4, 1, 3, 2, 5, 0], as the first element was removed.
+
+try {
+    list.removeAt(100)
+} catch (e: IndexOutOfBoundsException) {
+    // the list doesn't have 101 elements
+}
+// TODO: decide on it depending on whether we use asserts or checks
+```
+
+Here, we need to duplicate the removed element to show that there is no special
+behavior related to the value of the first element (that is, the other elements
+with the same value don't get removed).
+
+We also need `0` in the list to show that the `0` passed as the argument is
+unrelated to it.
+
+### Overrides of functions
+
+An override of a function must have its own usage sample if the types are
+different: either the covariant types are narrower or the contravariant types
+are wider.
+
+```kotlin
+val myFlow = MutableStateFlow(0)
+myFlow.collect {
+  // process the input
+} // can only exit with an exception
+throw IllegalStateException("This code will never be executed")
+```
+
+### Classes
+
+A class consists of three parts:
+
+* How its objects are obtained;
+* How its objects are used;
+* What other entities are available in its namespace (objects, functions, etc.).
+
+It is recommended providing top-level overview in the documentation for each of the aspect, ideally with the help of inline examples
+#### Construction
+
+The examples for constructors should be provided in the top-level documentation
+for the class and follow the same principles as the documentation of functions.
+
+```kotlin
+val parentJob = Job()
+// parentJob.children.toList() == listOf<Job>()
+val childJob = Job(parentJob)
+// parentJob.children.toList() == listOf<Job>(childJob)
+// childJob.parent == parentJob
+```
+
+Here, we must demonstrate both the known properties of the output and the
+mutating behavior of the constructor.
+
+#### Usage
+
+According to the guidelines, extension functions should typically encode the
+"extra" behavior, whereas member functions encode the core behavior.
+
+The top-level documentation should include an overview of the core behavior
+(typically, of the member functions), with simplified examples of each to
+quickly show one what behaviors are available. In these brief overviews,
+the rules for providing samples to functions don't need to be followed;
+instead, the sample must only demonstrate how *some* inputs and *some* forms of
+the output are linked. Edge cases and uncommon paths can be ignored.
+However, the rough idea of what the function does must still be deducible.
+
```

Comment:
#### Properties and fields

Properties and fields should be treated just like functions that accept the
class containing them as an argument; so, `val size: Int` of class `List`
should be accompanied by the samples that would be written for
`fun List<*>.size(): Int` according to the rules above:

```kotlin
check(listOf("a", "b", "c", "d").size == 4)
```

Mutable properties that have non-trivial setter behavior must also be treated
as functions that perform an action:

```kotlin
val input = "Mon, 30 Jul 2008 11:05:30 GMT"
val parsed = DateTimeComponents.Formats.RFC_1123.parse(input)
check(parsed.monthNumber == 7)
check(parsed.month == Month.JULY)
parsed.month = Month.JUNE
check(parsed.monthNumber == 6)
check(parsed.month == Month.JUNE)
```

Reply:
I've adjusted your recommendation in a separate commit, PTAL.
Rationale:

* Got rid of the notion of `field`. These are non-existing (except for things like `@JvmField`)
* Got rid of a trivial example -- it both makes the text a little bit more compact without the loss of the general idea and no longer promotes a very basic and trivial sample (trivial samples might be okay, but should be picked with care)

Reply:
Sorry, I haven't pushed it. I've meant the example with `size`. PTAL

Reply:
Yeah, I, too, meant the example with `size`: I think that, for how clear `size` inherently is, a one-line example is sufficient.

Here's a longer sample for the same idea, if by trivial you mean that it's too short:
```kotlin
check(LocalDate(2024, 1, 1).dayOfYear == 1)
check(LocalDate(2024, 1, 5).dayOfYear == 5)
check(LocalDate(2024, 2, 4).dayOfYear == 35)
check(LocalDate(2024, 12, 31).dayOfYear == 366)
```